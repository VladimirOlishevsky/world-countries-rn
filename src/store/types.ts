export interface ICountry {
    name: string,
    capital: string,
    region: string,
    subregion: string,
    population: number,
    latlng: number[],
    currencies: ICurrencies[],
    languages: ILanguages[],
    flag: string,
    alpha2Code: string,
    topLevelDomain: string[],
    callingCodes: string[]
}

export interface ICurrencies {
    name: string,
    symbol: string
}

export interface ILanguages {
    name: string,
    nativeName: string
}

export interface IRegionName {
    Name: string,
    FlagPng: string
    Alpha2Code: string
}