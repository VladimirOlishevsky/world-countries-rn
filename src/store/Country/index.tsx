import { action, runInAction, makeAutoObservable } from 'mobx';
import { ICountry, ICurrencies, ILanguages } from '../types';

type IObjectKeys = {
    [key: string]: string | number |  ICurrencies[] | ILanguages[];
  }

export interface ICountryStore2Obj extends IObjectKeys { 
    Capital: string,
    Region: string,
    Subregion: string,
    Population: number,
    Currencies: ICurrencies[],
    Languages: ILanguages[],
    Domain: string,
    PhoneCode: string,
}


export class Country {

    name = ''
    capital = ''
    region = ''
    subregion = ''
    population = 0
    latlng: number[] = []
    currencies: ICurrencies[] = []
    languages: ILanguages[] = []
    flag = ''
    alpha2Code = ''
    topLevelDomain: string[] = []
    callingCodes: string[] = []

    // errorMessage = ''

    constructor() {
        makeAutoObservable(this, {
            fromApi: action,
            fetchCountryByCode: action,
            fetchCountryByName: action,
            // getKeyByValue: action,
        });
    }
    async fetchCountryByCode(alphaCode: string) {
        // this.errorMessage = ''
        try {
            const response = await fetch(`https://restcountries.eu/rest/v2/alpha/${alphaCode}`)
            const data = await response.json();
            runInAction(() => {
                this.fromApi(data)
            });
        } catch (error) {
            runInAction(() => {
                console.log(error)
            })
        }
    }

    async fetchCountryByName(props: string) {
        // this.errorMessage = ''
        try {
            const response = await fetch(`https://restcountries.eu/rest/v2/name/${props}?fullText=true`)
            const data = await response.json();
            runInAction(() => {
                if (data.message) {
                    this.name = ''
                    return
                }
                this.fromApi(data[0])
            });
        } catch (error) {
            runInAction(() => {
                console.log(error)
            })
        }
    }

    get store2Obj(): ICountryStore2Obj {
        return {
            Capital: this.capital,
            Region: this.region,
            Subregion: this.subregion,
            Population: this.population,
            Currencies: this.currencies,
            Languages: this.languages,
            Domain: this.topLevelDomain[0],
            PhoneCode: this.callingCodes[0]
        }
    }

    fromApi(data: ICountry) {
        this.name = data.name;
        this.capital = data.capital;
        this.region = data.region;
        this.subregion = data.subregion;
        this.population = data.population;
        this.latlng = data.latlng;
        this.currencies = data.currencies
        this.languages = data.languages;
        this.flag = `https://flagcdn.com/w640/${data.alpha2Code.toLowerCase()}.png`;
        this.alpha2Code = data.alpha2Code;
        this.topLevelDomain = data.topLevelDomain
        this.callingCodes = data.callingCodes
    }
}