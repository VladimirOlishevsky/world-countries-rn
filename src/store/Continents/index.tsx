import { action, runInAction, makeAutoObservable } from 'mobx';
import { IRegionName } from 'store/types';
import { Region } from '../Region/index';

export class Continents {

    countries: IRegionName[] = [];

    constructor() {
        makeAutoObservable(this, {
            workerAfterFetch: action,
            fetchRegions: action
        })
    }
    async fetchRegions(props: string) {
        this.countries = [];

        try {
            const response = await fetch(props)
            const data = await response.json();
            const adaptData = this.workerAfterFetch(data.Response)
            runInAction(() => {
                this.countries = adaptData;
            });
        } catch (error) {
            console.log(error)
        }
    }

    workerAfterFetch(data: IRegionName[]) {
      return data.map(el => new Region(el))
    }
}