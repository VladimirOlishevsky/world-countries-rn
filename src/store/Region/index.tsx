import { action, makeAutoObservable } from 'mobx';
import { IRegionName } from 'store/types';

export class Region {

    Name = ''
    FlagPng = ''
    Alpha2Code = ''

    constructor(data: IRegionName) {
        makeAutoObservable(this, {
            fromapi: action,
        });
        data && this.fromapi(data)
    }

    fromapi(data: IRegionName) {
        this.Name = data.Name;
        this.FlagPng = data.FlagPng;
        this.Alpha2Code = data.Alpha2Code
    }
}
