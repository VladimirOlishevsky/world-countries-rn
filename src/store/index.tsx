import { Country } from "./Country";
import { RegionalBlocks } from "./RegionalBlocks";
import { Continents } from "./Continents";


class RootStore {
    countryStore: Country;
    regionalBlocksStore: RegionalBlocks;
    continentsStore: Continents;

  constructor() {
    this.countryStore = new Country();
    this.regionalBlocksStore = new RegionalBlocks();
    this.continentsStore = new Continents();
  }
}

const rootStore = new RootStore();
export type RootStoreType = RootStore;
export const getRootStore = (): RootStoreType => rootStore;
