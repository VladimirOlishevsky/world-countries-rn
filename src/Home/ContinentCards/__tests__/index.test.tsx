import React from "react";
import { create } from "react-test-renderer";
import { render, fireEvent } from '@testing-library/react-native';
import { ContinentCards } from "..";
import fetch, { enableFetchMocks } from 'jest-fetch-mock';
import { Continents } from "store/Continents";
import { Region } from "store/Region";


enableFetchMocks();

const mockData = [
  {
    Name: 'England',
    FlagPng: '111',
    Alpha2Code: '22'
  },
  {
    Name: 'Spain',
    FlagPng: '111',
    Alpha2Code: '22'
  },
]

jest.mock('..', () => (
  { ContinentCards: 'ContinentCards' }));

jest.mock('config/config', () => (
  { continents: 'continents' }));
  
const mockSetOpenModalConfirm = jest.fn();

describe('ContinentCards', () => {
  it('render ContinentCards', () => {
    const tree = render(<ContinentCards />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  // it('ContinentCards 11', async () => {

  //   // fetchMock.mockResponseOnce(JSON.stringify(mockData));
  //   const tree = render(<ContinentCards />);

  //   const page = tree.getByTestId('Europe');

  //   // fireEvent.press(page)
  //   // expect(fetch).toHaveBeenCalledTimes(1);

  //   const buttonBaseProps = page.props as {
  //     onClick: () => void
  //   };

  //   expect(mockSetOpenModalConfirm).not.toBeCalled();
  //   buttonBaseProps.onClick();
  //   // fireEvent.press(page)
  //   expect(mockSetOpenModalConfirm).toBeCalledWith(true);
  // });
});