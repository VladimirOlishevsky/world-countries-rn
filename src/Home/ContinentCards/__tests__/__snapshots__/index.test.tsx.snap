// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContinentCards render ContinentCards 1`] = `<ContinentCards />`;
